antsibull-docs: process routing metadata

felixfontein opened the pull request:
This PR adds code to antsibull-docs to process routing metadata (meta/runtime.yml for collections, and lib/ansible/config/ansible_builtin_runtime.yml for ansible-base/-core).

1. It removes duplicate module doc pages for redirects:
   - Before: https://docs.ansible.com/ansible/latest/collections/community/crypto/, see acme_account_facts and acme_account_info: the _facts module is a deprecated link to the _info module.
   - After: https://ansible.fontein.de/collections/community/crypto/index.html#plugins-in-community-crypto, the index only contains the non-deprecated acme_account_info, and https://ansible.fontein.de/collections/community/crypto/acme_account_facts_module.html still exists and informs about the deprecated redirect.
2. It adds stubs for all redirects; for example, in community.general 2.0.0, docker_container will be redirected to community.docker.docker_container.
    - The listing https://ansible.fontein.de/collections/community/general/ no longer contains docker_container
    - The file https://ansible.fontein.de/collections/community/general/docker_container_module.html still exists and contains information about the redirect.

This is similar to the stub pages created for the 2.9 docsite (see for example https://docs.ansible.com/ansible/2.9/modules/acme_account_facts_module.html; also there, acme_account_facts did not show up on https://docs.ansible.com/ansible/2.9/modules/list_of_crypto_modules.html or the global module index).

One "disadvantage" of the current form of the PR: it will also create redirect stub pages for all redirects in https://github.com/ansible/ansible/blob/devel/lib/ansible/config/ansible_builtin_runtime.yml. This is a huge amount of files, and I'm not sure whether we really want it. On the other hand, if ansible-core will ever move more content from itself to collections, this will make sure that when users switch in the version selector say from 2.10 to 4.0.0 on a plugin page where the plugin was moved in Ansible 4.0.0, the user sees that the module was moved.

This PR also makes switching between 2.10 and 3 friction-free: if a module has been moved somewhere else in Ansible 3, or has been removed, the user will directly see that (from the stub page they end up on). (For things already moved using the process in ansible-collections/overview#128, you can also switch back from 3 to 2.10 in the version selector and end up with a valid page.)

felixfontein commented:
The latest commit (9f28573df772db83ea949fc851e9a805bce06392) makes sure that only new stubs are created for ansible.builtin. For this, antsibull/data/ansible_2_10_routing.yml contains a list of all plugins that are redirected somewhere in ansible-base 2.10. This file never has to be adjusted, but we need it to be sure which redirects were added later (say, for plugins/modules that will be moved in ansible-core 2.11 or later).

felixfontein commented:
With this change:
```.sh
$ ls rst/collections/ansible/builtin/
add_host_module.rst
advanced_host_list_inventory.rst
apt_key_module.rst
apt_module.rst
apt_repository_module.rst
assemble_module.rst
assert_module.rst
async_status_module.rst
auto_inventory.rst
bigip_asm_policy_module.rst
bigip_facts_module.rst
bigip_gtm_facts_module.rst
blockinfile_module.rst
cmd_shell.rst
command_module.rst
config_lookup.rst
constructed_inventory.rst
copy_module.rst
cron_module.rst
csvfile_lookup.rst
debconf_module.rst
debug_module.rst
debug_strategy.rst
default_callback.rst
dict_lookup.rst
dnf_module.rst
dpkg_selections_module.rst
env_lookup.rst
expect_module.rst
fail_module.rst
fetch_module.rst
fileglob_lookup.rst
file_lookup.rst
file_module.rst
find_module.rst
first_found_lookup.rst
free_strategy.rst
gather_facts_module.rst
generator_inventory.rst
getent_module.rst
get_url_module.rst
git_module.rst
group_by_module.rst
group_module.rst
host_group_vars_vars.rst
host_list_inventory.rst
hostname_module.rst
host_pinned_strategy.rst
import_playbook_module.rst
import_role_module.rst
import_tasks_module.rst
include_module.rst
include_role_module.rst
include_tasks_module.rst
include_vars_module.rst
indexed_items_lookup.rst
index.rst
ini_inventory.rst
ini_lookup.rst
inventory_hostnames_lookup.rst
iptables_module.rst
items_lookup.rst
jsonfile_cache.rst
junit_callback.rst
known_hosts_module.rst
linear_strategy.rst
lineinfile_module.rst
lines_lookup.rst
list_lookup.rst
local_connection.rst
memory_cache.rst
meta_module.rst
minimal_callback.rst
nested_lookup.rst
oneline_callback.rst
package_facts_module.rst
package_module.rst
paramiko_ssh_connection.rst
password_lookup.rst
pause_module.rst
ping_module.rst
pipe_lookup.rst
pip_module.rst
powershell_shell.rst
psrp_connection.rst
random_choice_lookup.rst
raw_module.rst
reboot_module.rst
replace_module.rst
rpm_key_module.rst
runas_become.rst
script_inventory.rst
script_module.rst
sequence_lookup.rst
service_facts_module.rst
service_module.rst
set_fact_module.rst
set_stats_module.rst
setup_module.rst
shell_module.rst
sh_shell.rst
slurp_module.rst
ssh_connection.rst
stat_module.rst
su_become.rst
subelements_lookup.rst
subversion_module.rst
sudo_become.rst
systemd_module.rst
sysvinit_module.rst
tempfile_module.rst
template_lookup.rst
template_module.rst
together_lookup.rst
toml_inventory.rst
tree_callback.rst
unarchive_module.rst
unvault_lookup.rst
uri_module.rst
url_lookup.rst
user_module.rst
varnames_lookup.rst
vars_lookup.rst
wait_for_connection_module.rst
wait_for_module.rst
winrm_connection.rst
yaml_inventory.rst
yum_module.rst
yum_repository_module.rst
```
(https://ansible.fontein.de/collections/ansible/builtin/index.html)

gundalow commented:
Thank you for doing this. The shorter list on https://ansible.fontein.de/collections/community/crypto/index.html#plugins-in-community-crypto is a great improvement

https://ansible.fontein.de/collections/community/crypto/acme_account_facts_module.html
1. Wonder if we should break the paragraph into a bullet list. Feels slightly like a wall of text
2. Could we make the call to action clearer (maybe this should be the first point)
3. For redirects should `To use it in a playbook, specify` be removed, or maybe use the new name?

felixfontein commented:
@gundalow I've adjusted the template in c9ca790937c96608bc4adfaf4e6d3a6918990a71, see https://ansible.fontein.de/collections/community/crypto/acme_account_facts_module.html for the result.

> Could we make the call to action clearer (maybe this should be the first point)

I've moved it up.

> For redirects should To use it in a playbook, specify be removed, or maybe use the new name?

For now I've kept it for redirects, and removed it for deprecated redirects. Maybe it should say *"To use this redirect in a playbook, specify: `foo.bar.old_name`. To directly reference the new location, specify: `baz.bar.new_name`".* What do you think?

felixfontein commented:
Example of a tombstoned module: https://ansible.fontein.de/collections/community/general/docker_image_facts_module.html

abadger commented on antsibull/docs_parsing/routing.py:
```diff
@@ -0,0 +1,309 @@
+
+
+def add_symlink(collection_name: str, src_components: t.List[str], dest_components: t.List[str],
+                plugin_type_routing: t.Dict[str, t.Dict[str, t.Any]]) -> None:
```
As much as possible, I like to create functions rather than procedures that operate by side effect.  I think this one definitely can return its value rather than modifying `plugin_type_routing`.  I have a feeling that `add_symlinks()` can also return its value but I'm not as sure there.  returning the value would mean moving some logic to the higher level and I haven't quite decided if that would be a good thing or not.

Reasons for liking functions:
* Side effects are harder to keep in mind than return values.
* Unit testing is more straightforward for functions than procedures which operate by side effect.

felixfontein commented on antsibull/yaml.py:
```diff
@@ -0,0 +1,47 @@
+    with open(path, 'w') as stream:
+        dumper = _SafeDumper
+        dumper.ignore_aliases = lambda *args: True
+        yaml.dump(content, stream, default_flow_style=False, encoding='utf-8', Dumper=dumper)
```
Implemented in 76b6f3ee1174daa2be946a938093516181370a46.

felixfontein commented on antsibull/write_docs.py:
```diff
@@ -113,6 +113,74 @@ async def write_rst(collection_name: str, collection_meta: AnsibleCollectionMeta
     flog.debug('Leave')
 
 
+async def write_stub_rst(collection_name: str, collection_meta: AnsibleCollectionMetadata,
```
Implemented in b61d9f3225094a2a4612a9ee307a109e2032d00d.

felixfontein commented on antsibull/yaml.py:
```diff
@@ -0,0 +1,47 @@
+YAML handling.
+"""
+
+from typing import Any
```
Implemented in 8eea464bcc4d0b205d74a84823dafc494067a6c1.

felixfontein commented on antsibull/docs_parsing/routing.py:
```diff
@@ -0,0 +1,309 @@
+
+def find_stubs(plugin_info: t.MutableMapping[str, t.MutableMapping[str, t.Any]],
+               collection_routing: CollectionRoutingT
+               ) -> t.Mapping[str, t.Mapping[str, t.Mapping[str, t.Any]]]:
```
We did discuss it here: https://github.com/ansible-community/antsibull/pull/200#discussion_r529922217 and in IRC (on November 30th, just read up the logs). I'm not sure we really settled on something, I still think it's better to not let the implementation leak into the return value. But in the end it doesn't matter that much to me, I've adjusted the return types in 5ae211f6c34b7096fd91740efbe7cfdd61dbd5ab.

felixfontein commented on antsibull/docs_parsing/routing.py:
```diff
@@ -202,14 +203,19 @@ async def load_collection_routing(collection_name: str,
+    if collection_name == 'ansible.builtin':
+        # ansible-base/-core has a special directory structure we currently do not want
+        # (or need) to handle
+        return
```
This should return `plugin_routing_out`.

felixfontein commented on antsibull/docs_parsing/routing.py:
```diff
@@ -206,11 +220,23 @@ async def load_collection_routing(collection_name: str,
+
+        symlink_redirects = find_symlink_redirects(collection_name, plugin_type, directory_path)
+        for redirect_name, redirect_dst in symlink_redirects.items():
+            if 'redirect' not in plugin_type_routing[redirect_name]:
```
This will raise an `AttributeError` if `redirect_name not in plugin_type_routing`. That actually does happen for community.general. Adding
```.py
            if redirect_name not in plugin_type_routing:
                plugin_type_routing[redirect_name] = {}
```
before this line makes it work. Or at least a bit longer, until the code crashes in `find_flatmapping_short_long_maps` because `redirect = routing_data['redirect']` is not a string, but a dictionary with a key `redirect` :)

felixfontein replied:
Fixed in bac0aca263ba6aeb8074623e5cff17356b41befb.

felixfontein commented on antsibull/docs_parsing/routing.py:
```diff
@@ -59,35 +69,39 @@ def calculate_fqcns(basename, file_path, plugin_type, rel_path, path, collection
+                src_fqcn, dst_fqcn = calculate_plugin_fqcns(collection_name, src_basename,
+                                                            dst_basename, rel_path)
+
+                plugin_type_routing[src_fqcn]['redirect'] = dst_fqcn
```
This needs to be changed to
```.py
                 plugin_type_routing[src_fqcn] = dst_fqcn
```
and the type of `plugin_type_routing` to `Dict[str, str]` to fix the bug mentioned above.

With both these changes, it works as expected. I'll push a commit to fix it.

felixfontein replied:
Fixed in bac0aca263ba6aeb8074623e5cff17356b41befb.

felixfontein commented on antsibull/docs_parsing/routing.py:
```diff
@@ -0,0 +1,326 @@
+                    # redirected one are generated from the same plugin filename, then we can
+                    # remove this plugin's docs and generate a redirect stub instead.
+                    if compare_all_but(plugin_record['doc'], plugin_map[destination]['doc'],
+                                       'filename'):
```
```suggestion
                                       ['filename']):
```
Otherwise it ignores the keys `f`, `i`, `l`, ...

abadger commented on antsibull/docs_parsing/routing.py:
```diff
@@ -0,0 +1,326 @@
+            if isinstance(date, datetime.date):
+                date = date.isoformat()
+            plugin_record[tlkey]['removal_date'] = date
+    return plugin_record
```
I don't really like that this function both returns plugin_record and mutates plugin_record.  But I'm not sure if there's an easy way to fix that.  It would be preferable as an API if it returned plugin_record but that would require copying plugin_record when it's not needed.  OTOH, mutating the plugin_record without returning it here won't fit in with the  way we're using it in load_collection_routing.

felixfontein commented:
@abadger thanks for your work on this!
